lib/Utils/Graph/Graph.h

```diff
@@ -154,6 +157,160 @@ class Graph {
   std::map<V, std::set<V>> inEdges;
 };
 
+// An undirected graph data structure.
+//
+// Parameter `V` is the vertex type, which is expected to be cheap to copy.
+template <typename V>
+class UndirectedGraph {
+ public:
+  // Adds a vertex to the graph
+  void addVertex(V vertex) { vertices.insert(vertex); }
+
+  // Adds an edge between `source` and `target`. Returns false if either the
+  // source or target is not a previously inserted vertex, and returns true
+  // otherwise. The graph is unchanged if false is returned.
+  bool addEdge(V source, V target) {
+    if (vertices.count(source) == 0 || vertices.count(target) == 0) {
+      return false;
+    }
+    edges[source].insert(target);
+    edges[target].insert(source);
+    return true;
+  }
+
+  // Returns true iff the given vertex has previously been added to the graph
+  // using `AddVertex`.
+  bool contains(V vertex) const { return vertices.count(vertex) > 0; }
+
+  bool empty() const { return vertices.empty(); }
+
+  const std::set<V>& getVertices() const { return vertices; }
+
+  // Returns the edges incident to the given vertex.
+  std::vector<V> edgesIncidentTo(V vertex) const {
+    if (vertices.count(vertex)) {
+      std::vector<V> result(edges.at(vertex).begin(), edges.at(vertex).end());
+      // Note: The vertices are sorted to ensure determinism in the output.
+      std::sort(result.begin(), result.end());
+      return result;
+    }
+    return {};
+  }
+
+ private:
+  std::set<V> vertices;
+  std::map<V, std::set<V>> edges;
+};
+
+/// The greedy "DSatur" graph coloring algorithm
+/// Cf. https://en.wikipedia.org/wiki/DSatur
+template <typename V>
+class GreedyGraphColoring {
+ public:
+  GreedyGraphColoring() = default;
+  ~GreedyGraphColoring() = default;
+
+  std::unordered_map<V, int> color(const UndirectedGraph<V>& graph) {
+    colors.clear();
+    neighborColors.clear();
+    vertexSaturations.clear();
+
+    for (const auto& vertex : graph.getVertices()) {
+      neighborColors[vertex] = std::unordered_set<int>();
+      updateSaturationDegree(graph, vertex);
+    }
+
+    while (!queue.empty()) {
+      auto current = queue.top();
+      queue.pop();
+
+      // Skip if vertex is already colored or info is outdated. Could avoid
+      // having "stale" info by using a std::set instead of a priority queue,
+      // but then that would incur log(n) lookups and log(n) updates. Probably
```

Comment:
Thanks for this comment! I was exactly wondering whether pushing the same vertices on the stack when it's neighbor is colored (from `updateNeighborSaturation` pushing onto the queue) was an issue. 

I think you could also erase it from the queue before you update the neighborColors map on line 293, but maybe only if it's performance reason related, since it would be nice then to move the queue update out of the helper in `updateSaturationDegree`.

Reply:
Since erasing has log-time lookups, I feel it would be faster overall for sparse graphs to wait until they are popped naturally and skip them. You would do this operation a constant number of times (degree(v) times is constant for sparse graphs) while it could be log(num_vertices) deletion even for sparse graphs.